### ISSUE TITLE
Opening Indices blocks all indexing threads.

### ISSUE THREAD

**Issue by ryanhadley** [elastic/elasticsearch#17744]
**Elasticsearch version**: 2.2.0

**JVM version**: Java(TM) SE Runtime Environment (build 1.8.0_71-b15)

**OS version**: CentOS release 6.6

**Description of the problem including expected versus actual behavior**:
Opening indices can cause indexing tasks to time out while processing put-mapping events.

We have about 200,000 indices on our cluster, but typically only have 100-300 indices open. We close them after a period of inactivity and re-open them the next time they're needed. If a large number of requests come in to open closed indices, this causes the put-mapping events generated by bulk indexing threads to hit the 30 second time out and fail.

From watching the pending_tasks list, I suspect (but could be very wrong here) that it's because the open requests insert cluster state tasks of 'shard-started' with URGENT priority. The indexing jobs insert cluster state tasks of 'put-mapping' with HIGH priority. It appears that the URGENT ones always process before the HIGH ones, effectively blocking any indexing jobs.

**Steps to reproduce**:
1. Create a bunch of indices. 1000's
2. Close all of them
3. Start a loop opening those closed indices one at a time
4. Start indexing a new index
5. Your indexing job will now fail with a timeout processing cluster event put-mapping

**Provide logs (if relevant)**:
ProcessClusterEventTimeoutException[failed to process cluster event (put-mapping [response]) within 30s]
        at org.elasticsearch.cluster.service.InternalClusterService$2$1.run(InternalClusterService.java:343)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)

**Comment by clintongormley** [comment]
@bleskes could you take a look at this please?

**Comment by bleskes** [comment]
Thx @ryanhadley . The reason why shard started messages are processed with high priority is that they are needed to process index and search operations and thus have a higher impact that a put mapping which may block some operations. That said they are batched and should be light, so I wonder why they talk long.  Also, it may be that something else is block the put mapping requests.

Since you say you looked at the pending task list, can you maybe share it? I would also be interested to see the output of the hot threads API of the master node while it processes this open index/put mapping requests

**Comment by ryanhadley** [comment]
@bleskes, thank you! I'm trying to get you this additional data (can't believe I didn't save it from before... oops) but it might take me a bit. We worked around the problem on our Prod cluster by pooling open requests in a queue and only running the open requests once every 10 seconds. I have to get a test cluster up and running with enough large indices to close and then open to recreate this.

> That said they are batched and should be light, so I wonder why they talk long.

They are definitely light and complete very fast. The open request typically takes less than a second per index. The only time it causes issues is when there are many open requests in a row. Which apparently our customers like to do in the early mornings when they show up to work and want to check their reports for new data.

**Comment by ryanhadley** [comment]
OK, I got enough data in to a test cluster I spun up to recreate the issue.

This is a 4 server cluster and I got about 500 indices in to it of decent size (50,000 - 100,000 average document count per index). I close all but 200 of the indexes and then ran this to start opening the closed indices one at a time:

for sindex in `curl 'http://localhost:9200/_cluster/state/blocks?pretty' 2>/dev/null | grep "survey-" | awk '{print $1}' | sed -e 's/"//g'`; do echo $sindex; curl -XPOST "http://localhost:9200/$sindex/_open"; done

This was MUCH more stressful on the cluster than what our customers traffic would drive, but it results in the same put-mapping timeout errors while indexing and the same constant listing of URGENT priority tasks in the pending_tasks list while the put-mapping timeouts happen.

Attached are the results of pending_tasks and hot_threads.

[pending_tasks.json.txt](https://github.com/elastic/elasticsearch/files/219885/pending_tasks.json.txt)
[hot_threads.txt](https://github.com/elastic/elasticsearch/files/219887/hot_threads.txt)

**Comment by dragonrider23** [comment]
I seem to have come across the same issue. Windows 10 running a single node during the hands-on training. Elasticsearch hung with no indication anything was wrong. When I stopped it, it output the following:

```
[2016-05-03 14:28:53,418][INFO ][cluster.metadata         ] [web-1] [.kibana] create_mapping [visualization]
[2016-05-03 14:38:07,643][INFO ][cluster.metadata         ] [web-1] [.kibana] update_mapping [index-pattern]
[2016-05-03 15:03:36,649][WARN ][rest.suppressed          ] /.kibana Params: {index=.kibana}
ProcessClusterEventTimeoutException[failed to process cluster event (delete-index [.kibana]) within 30s]
        at org.elasticsearch.cluster.service.InternalClusterService$2$1.run(InternalClusterService.java:349)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
[2016-05-03 15:03:30,643][WARN ][rest.suppressed          ] /.kibana Params: {index=.kibana}
ProcessClusterEventTimeoutException[failed to process cluster event (delete-index [.kibana]) within 30s]
        at org.elasticsearch.cluster.service.InternalClusterService$2$1.run(InternalClusterService.java:349)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
[2016-05-03 14:58:56,366][DEBUG][action.admin.indices.mapping.put] [web-1] failed to put mappings on indices [[.kibana]]
, type [dashboard]
```

**Comment by bleskes** [comment]
@ryanhadley sorry for taking so long, but looking at the dumps you sent, it seemes the master is busy processing the refresh mapping commands it gets:

```
   79.7% (398.6ms out of 500ms) cpu usage by thread 'elasticsearch[ip-10-0-2-97-indexing-01][clusterService#updateTask][T#1]'
     2/10 snapshots sharing following 12 elements
       org.elasticsearch.index.mapper.MapperService.addMappers(MapperService.java:425)
       org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:318)
       org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:272)
       org.elasticsearch.cluster.metadata.MetaDataMappingService.executeRefresh(MetaDataMappingService.java:140)
       org.elasticsearch.cluster.metadata.MetaDataMappingService$RefreshTaskExecutor.execute(MetaDataMappingService.java:79)

```

Can you put your cluster on debug logging and shard the logs? I'm looking for messages that are produced by this line:

```
logger.debug("[{}] parsed mapping [{}], and got different sources\noriginal:\n{}\nparsed:\n{}", index, mappingType, mappingSource, mapperService.documentMapper(mappingType).mappingSource());
```

**Comment by colings86** [comment]
No further feedback

**Comment by ryanhadley** [comment]
Sorry I didn't get back to you. But we discovered that we were just expecting things from Elasticsearch that we shouldn't have been expecting. Our cluster state was so large that things that should happen quickly were not happening quick enough. We solved this solution by drastically reducing our cluster state size with more condensed mapping and less indexes total.

The cluster state was so large. So large. We did bad things.
